spanner/integration_test.go

```diff
@@ -2476,6 +2476,366 @@ func TestIntegration_TransactionRunner(t *testing.T) {
+		iter = client2.Single().Read(ctx, "Singers", AllKeys(), singerColumns)
+		defer iter.Stop()
+
+		_, err = iter.Next()
+		if err == nil {
+			t.Fatal("expected err, got nil")
+		}
+		if msg, ok := matchError(err, codes.PermissionDenied, test.errMsg); !ok {
+			t.Fatal(msg)
+		}
+	}
+}
+
+func TestIntegration_DMLWithRoles(t *testing.T) {
+	t.Parallel()
+
+	// Set up testing environment.
+	var (
+		client, client2 *Client
+		err             error
+	)
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Minute)
+	defer cancel()
+	stmts := []string{
+		`CREATE TABLE Singers (
+				SingerId	INT64 NOT NULL,
+				FirstName	STRING(1024),
+				LastName	STRING(1024),
+				SingerInfo	BYTES(MAX)
+			) PRIMARY KEY (SingerId)`,
+		`CREATE ROLE singers_writer`,
+		`CREATE ROLE singers_unauthorized`,
+		`GRANT SELECT(SingerId), UPDATE(FirstName, LastName) ON TABLE Singers TO ROLE singers_writer`,
+		`GRANT SELECT(SingerId), UPDATE(FirstName) ON TABLE Singers TO ROLE singers_unauthorized`,
+	}
+	client, dbPath, cleanup := prepareIntegrationTest(ctx, t, DefaultSessionPoolConfig, stmts)
+	defer cleanup()
+
+	singerColumns := []string{"SingerId", "FirstName", "LastName"}
+	var ms = []*Mutation{
+		InsertOrUpdate("Singers", singerColumns, []interface{}{1, "Marc", "Richards"}),
+	}
+	if _, err := client.Apply(ctx, ms, ApplyAtLeastOnce()); err != nil {
+		t.Fatalf("Could not insert rows to table. Got error %v", err)
+	}
+	updateStmt := Statement{SQL: `UPDATE Singers SET FirstName = "Mark", LastName = "Richards" WHERE SingerId = 1`}
+
+	// A request with sufficient privileges should update the row
+	for _, dbRole := range []string{
+		"",
+		"singers_writer",
+	} {
+		if client2, err = createClientWithRole(ctx, dbPath, SessionPoolConfig{}, dbRole); err != nil {
+			t.Fatal(err)
+		}
+		defer client2.Close()
+		_, err = client2.ReadWriteTransaction(ctx, func(ctx context.Context, txn *ReadWriteTransaction) error {
+			_, err := txn.Update(ctx, updateStmt)
+			return err
+		})
+		if err != nil {
+			t.Fatalf("Could not update row. Got error %v", err)
+		}
+	}
+
+	// A request with insufficient privileges should return permission denied
+	for _, test := range []struct {
+		dbRole string
+		errMsg string
+	}{
+		{
+			"singers_unauthorized",
+			"Role singers_unauthorized does not have required privileges on table Singers.",
+		},
+		{
+			"nonexistent",
+			"Role not found: nonexistent.",
+		},
+	} {
+		if client2, err = createClientWithRole(ctx, dbPath, SessionPoolConfig{}, test.dbRole); err != nil {
+			t.Fatal(err)
+		}
+		defer client2.Close()
+		_, err = client2.ReadWriteTransaction(ctx, func(ctx context.Context, txn *ReadWriteTransaction) error {
+			_, err := txn.Update(ctx, updateStmt)
+			return err
+		})
+
+		if err == nil {
+			t.Fatal("expected err, got nil")
+		}
+		if msg, ok := matchError(err, codes.PermissionDenied, test.errMsg); !ok {
+			t.Fatal(msg)
+		}
+	}
+}
+
+func TestIntegration_MutationWithRoles(t *testing.T) {
```

Comment:
Could you also add a test for Insert/Delete mutation?

Reply:
Added these to `TestIntegration_DMLWithRoles` and `TestIntegration_MutationWithRoles`.